Adhere to async and reactive APIs (#1)
dwdougherty/redis-doc: docs/connect/clients/lettuce.md

### Asynchronous connection

```java
package org.example;
import java.util.*;
import java.util.concurrent.ExecutionException;

import io.lettuce.core.*;
import io.lettuce.core.api.async.RedisAsyncCommands;
import io.lettuce.core.api.StatefulRedisConnection;

public class Async {
  public static void main(String[] args) {
    RedisClient redisClient = RedisClient.create("redis://localhost:6379");

    try (StatefulRedisConnection<String, String> connection = redisClient.connect()) {
      RedisAsyncCommands<String, String> asyncCommands = connection.async();

>       // Asynchronously store & retrieve a simple string

Reviewer: Thank you @tishun. I'll make a note of this for the next iteration.

      asyncCommands.set("foo", "bar").get();
      System.out.println(asyncCommands.get("foo").get()); // prints bar

>       // Asynchronously store key-value pairs in a hash directly

Reviewer 2: What is the benefit of this specific example?
Perhaps we can show here some other feature that makes sense in an async approach, e.g. handling errors?

Reviewer: The other client guides all use mostly the same examples, so that's why this is here.

      hash.put("surname", "Smith");
      hash.put("company", "Redis");
      hash.put("age", "29");
      asyncCommands.hset("user-session:123", hash).get();

      System.out.println(asyncCommands.hgetall("user-session:123").get());
      // Prints: {name=John, surname=Smith, company=Redis, age=29}
    } catch (ExecutionException | InterruptedException e) {
      throw new RuntimeException(e);
    } finally {
      redisClient.shutdown();
    }
  }
}
```

Learn more about asynchronous Lettuce API in [the reference guide](https://lettuce.io/core/release/reference/index.html#asynchronous-api).

### Reactive connection

```java
package org.example;
import java.util.*;
import io.lettuce.core.*;
import io.lettuce.core.api.reactive.RedisReactiveCommands;
import io.lettuce.core.api.StatefulRedisConnection;

public class Main {
  public static void main(String[] args) {
    RedisClient redisClient = RedisClient.create("redis://localhost:6379");

    try (StatefulRedisConnection<String, String> connection = redisClient.connect()) {
      RedisReactiveCommands<String, String> reactiveCommands = connection.reactive();

      // Reactively store & retrieve a simple string

>       reactiveCommands.set("foo", "bar").block();

Reviewer 2: To fully take advantage of the async approach we should try and block once for this, similar to the example I added for the Async/Future approach

      reactiveCommands.get("foo").doOnNext(System.out::println).block(); // prints bar

      // Reactively store key-value pairs in a hash directly
      Map<String, String> hash = new HashMap<>();
      hash.put("name", "John");
      hash.put("surname", "Smith");
      hash.put("company", "Redis");
      hash.put("age", "29");

>       reactiveCommands.hset("user-session:124", hash).then(

Reviewer 2: I think we should handle all the error cases, just to show the best practices.

Applies to all examples.

```java
RedisURI redisUri = RedisURI.Builder.redis("localhost")
                                 .withSsl(true)
                                 .withPassword("secret!") // use your Redis password
                                 .build();

RedisClient client = RedisClient.create(redisUri);
```

## Connection Management in Lettuce

Lettuce uses `ClientResources` for efficient management of shared resources like event loop groups and thread pools.
For connection pooling, Lettuce leverages `RedisClient` or `RedisClusterClient`, which can handle multiple concurrent connections efficiently.

A typical approach with Lettuce is to create a single `RedisClient` instance and reuse it to establish connections to your Redis server(s).

> These connections are multiplexed; that is, multiple commands can be run concurrently over a single or a small set of connections, making explicit pooling less critical than in blocking clients like Jedis.

Reviewer 3: Also, it is not _less critical_ anymore when you introduce pooling.

> Lettuce comes with an asynchronous, non-blocking pool implementation to be used with Lettuces asynchronous connection methods. It does not require additional dependencies.

Reviewer 3: > Lettuce comes with an asynchronous, non-blocking pool implementation ...

NO, IT DOESN'T.
The pool implementation is actually of commons-pool2 library.
Lettuce provides _wrapper_ classes (e.g. `BoundedPoolConfig`) for commons-pool2 config classes.

> It does not require additional dependencies.

IT DOES. E.g. `BoundedPoolConfig.create()` below depends on commons-pool2 library. The statement will fail without that library.

```suggestion
Lettuce provides pool config to be used with Lettuces asynchronous connection methods.
```

```java
package org.example;
import io.lettuce.core.RedisClient;
import io.lettuce.core.RedisURI;
import io.lettuce.core.TransactionResult;
import io.lettuce.core.api.StatefulRedisConnection;
import io.lettuce.core.api.async.RedisAsyncCommands;
import io.lettuce.core.codec.StringCodec;
import io.lettuce.core.support.*;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public class Pool {
  public static void main(String[] args) {
    RedisClient client = RedisClient.create();

    String host = "localhost";
    int port = 6379;

    CompletionStage<BoundedAsyncPool<StatefulRedisConnection<String, String>>> poolFuture = AsyncConnectionPoolSupport.createBoundedObjectPoolAsync(
        () -> client.connectAsync(StringCodec.UTF8, RedisURI.create(host, port)),

>         BoundedPoolConfig.create());

Reviewer 3: It is worth considering a detailed config; e.g. `BoundedPoolConfig.builder().maxTotal(100).maxIdle(20).minIdle(5).build()`.

Reviewer: @sazzad16 I'm not a Java programmer, as you could likely tell. Would I just replace the line:

BoundedPoolConfig.create());

with the line you provided in your comment?

Reviewer: Also, thank you very much for the other comments! Much appreciated.

Reviewer 2: Can't we just use the example in https://github.com/lettuce-io/lettuce-core/wiki/Connection-Pooling ?

Reviewer 3: @dwdougherty Yes, you can simply replace `BoundedPoolConfig.create()` with line/statement in my comment.
